rust/experimental/query_engine/kql-parser/src/scalar_expression.rs

```diff
@@ -13,28 +13,50 @@ pub(crate) fn parse_scalar_expression(
-        Rule::real_expression => Ok(ScalarExpression::Static(parse_real_expression(
-            scalar_rule,
-        )?)),
-        Rule::datetime_expression => Ok(ScalarExpression::Static(parse_datetime_expression(
-            scalar_rule,
-        )?)),
-        Rule::logical_expression => Ok(ScalarExpression::Logical(
-            parse_logical_expression(scalar_rule, state)?.into(),
-        )),
-        Rule::true_literal | Rule::false_literal => Ok(ScalarExpression::Static(
-            parse_standard_bool_literal(scalar_rule),
-        )),
-        Rule::double_literal => Ok(ScalarExpression::Static(parse_double_literal(scalar_rule)?)),
-        Rule::integer_literal => Ok(ScalarExpression::Static(parse_integer_literal(
-            scalar_rule,
-        )?)),
-        Rule::string_literal => Ok(ScalarExpression::Static(parse_string_literal(scalar_rule))),
-        Rule::accessor_expression => parse_accessor_expression(scalar_rule, state),
-        Rule::conditional_expression => parse_conditional_expression(scalar_rule, state),
-        Rule::scalar_expression => parse_scalar_expression(scalar_rule, state),
+    let scalar = match scalar_rule.as_rule() {
+        Rule::real_expression => ScalarExpression::Static(parse_real_expression(scalar_rule)?),
+        Rule::datetime_expression => {
+            ScalarExpression::Static(parse_datetime_expression(scalar_rule)?)
+        }
+        Rule::true_literal | Rule::false_literal => {
+            ScalarExpression::Static(parse_standard_bool_literal(scalar_rule))
+        }
+        Rule::double_literal => ScalarExpression::Static(parse_double_literal(scalar_rule)?),
+        Rule::integer_literal => ScalarExpression::Static(parse_integer_literal(scalar_rule)?),
+        Rule::string_literal => ScalarExpression::Static(parse_string_literal(scalar_rule)),
+        Rule::accessor_expression => parse_accessor_expression(scalar_rule, state)?,
+        Rule::logical_expression => {
+            let l = parse_logical_expression(scalar_rule, state)?;
+
+            if let LogicalExpression::Scalar(s) = l {
+                s
+            } else {
+                ScalarExpression::Logical(l.into())
+            }
+        }
+        Rule::conditional_expression => parse_conditional_expression(scalar_rule, state)?,
+        Rule::scalar_expression => parse_scalar_expression(scalar_rule, state)?,
         _ => panic!("Unexpected rule in scalar_expression: {}", scalar_rule),
+    };
+
+    if matches!(&scalar, ScalarExpression::Static(_)) {
+        return Ok(scalar);
+    }
+
+    // Note: What this branch does is test if the scalar being returned resolves
+    // to a static value. If it does the whole expression is folded/replaced
+    // with the resolved static value. This generally shrinks the expression
+    // tree and makes it faster to execute.
+    let static_result = scalar.try_resolve_static();
+    if let Err(e) = static_result {
+        Err(ParserError::SyntaxError(
```

Comment:
Note: I tried to get coverage here but it isn't currently possible to make this get hit. I'll get it on my next PR where I am adding folding inside of accessor expressions. I think it will be possible then.